libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtilsTest.kt

```diff
@@ -512,6 +543,80 @@ class MapboxRouteLineUtilsTest {
         verify(exactly = 0) { style.addStyleSource(any(), any()) }
     }
 
+    @Test
+    fun initializeLayers_whenLayersAreNotInitialized() {
```

Comment:
What do you think about checking gl-native usage in general, without duplicating internal implementation?

```
Index: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtilsTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtilsTest.kt b/libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtilsTest.kt
--- a/libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtilsTest.kt	(revision 4b53565e6661791eb05d79a80d576a41563c9641)
+++ b/libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtilsTest.kt	(date 1679312787023)
@@ -12,6 +12,7 @@
 import com.mapbox.maps.StyleObjectInfo
 import com.mapbox.maps.extension.style.layers.Layer
 import com.mapbox.maps.extension.style.layers.addPersistentLayer
+import com.mapbox.maps.extension.style.layers.generated.LineLayer
 import com.mapbox.maps.extension.style.layers.getLayer
 import com.mapbox.maps.extension.style.layers.properties.generated.Visibility
 import com.mapbox.navigation.base.internal.NativeRouteParserWrapper
@@ -550,67 +551,46 @@
                 .lineDepthOcclusionFactor(0.85)
                 .displayRestrictedRoadSections(true)
                 .build()
-            val lineLayerIds = listOf(
-                LAYER_GROUP_1_TRAIL_CASING,
-                LAYER_GROUP_1_TRAIL,
-                LAYER_GROUP_1_CASING,
-                LAYER_GROUP_1_MAIN,
-                LAYER_GROUP_1_TRAFFIC,
-                LAYER_GROUP_1_RESTRICTED,
-                LAYER_GROUP_2_TRAIL_CASING,
-                LAYER_GROUP_2_TRAIL,
-                LAYER_GROUP_2_CASING,
-                LAYER_GROUP_2_MAIN,
-                LAYER_GROUP_2_TRAFFIC,
-                LAYER_GROUP_2_RESTRICTED,
-                LAYER_GROUP_3_TRAIL_CASING,
-                LAYER_GROUP_3_TRAIL,
-                LAYER_GROUP_3_CASING,
-                LAYER_GROUP_3_MAIN,
-                LAYER_GROUP_3_TRAFFIC,
-                LAYER_GROUP_3_RESTRICTED,
-                MASKING_LAYER_TRAIL_CASING,
-                MASKING_LAYER_TRAIL,
-                MASKING_LAYER_CASING,
-                MASKING_LAYER_MAIN,
-                MASKING_LAYER_TRAFFIC,
-                MASKING_LAYER_RESTRICTED,
-            )
+            val initializedLineLayersIds = mutableSetOf<String>()
+            val sourcesUsedByLayers = mutableSetOf<String>()
+            val initializedSourceIds = mutableSetOf<String>()
             val style = mockk<Style>(relaxed = true) {
-                every { addPersistentLayer(any(), any()) } answers {}
                 every { styleLayers } returns listOf()
 
-                every { styleSourceExists(LAYER_GROUP_1_SOURCE_ID) } returns false
-                every { styleSourceExists(LAYER_GROUP_2_SOURCE_ID) } returns false
-                every { styleSourceExists(LAYER_GROUP_3_SOURCE_ID) } returns false
-                every { styleSourceExists(BOTTOM_LEVEL_ROUTE_LINE_LAYER_ID) } returns false
-                every { styleSourceExists(TOP_LEVEL_ROUTE_LINE_LAYER_ID) } returns false
-                every { styleSourceExists(WAYPOINT_SOURCE_ID) } returns false
+                every { styleSourceExists(any()) } answers {
+                    initializedSourceIds.contains(firstArg())
+                }
+                every { addStyleSource(any(), any()) } answers {
+                    initializedSourceIds.add(firstArg())
+                    ExpectedFactory.createNone()
+                }
 
-                lineLayerIds.forEach {
-                    every { styleSourceExists(it) } returns false
+                every { addPersistentLayer(any(), any()) } answers {
+                    val layer = secondArg<Layer>()
+                    if (layer is LineLayer) {
+                        initializedLineLayersIds.add(layer.layerId)
+                    }
+
+                    val sourceId = Layer::class.java.declaredMethods
+                        .first { it.name.startsWith("getInternalSourceId") }
+                        .invoke(layer) as String?
+                    if (sourceId != null) {
+                        sourcesUsedByLayers.add(sourceId)
+                    }
                 }
             }
 
             MapboxRouteLineUtils.initializeLayers(style, options)
 
-            verify(exactly = 1) {
-                style.addStyleSource(LAYER_GROUP_1_SOURCE_ID, any())
-                style.addStyleSource(LAYER_GROUP_2_SOURCE_ID, any())
-                style.addStyleSource(LAYER_GROUP_3_SOURCE_ID, any())
-                style.addPersistentLayer(
-                    match { it.layerId == TOP_LEVEL_ROUTE_LINE_LAYER_ID },
-                    any()
-                )
-                style.addPersistentLayer(
-                    match { it.layerId == BOTTOM_LEVEL_ROUTE_LINE_LAYER_ID },
-                    any()
-                )
-                style.addStyleSource(WAYPOINT_SOURCE_ID, any())
+            sourcesUsedByLayers.forEach { sourceId ->
+                verify(exactly = 1) {
+                    style.addStyleSource(sourceId, any())
+                }
             }
-            lineLayerIds.forEach { layerId ->
+
+            initializedLineLayersIds
+                .forEach { layerId ->
                 verify(exactly = 1) {
-                    style.addPersistentLayer(match { it.layerId == layerId }, any())
                     style.setStyleLayerProperty(layerId, "line-depth-occlusion-factor", Value(0.85))
                 }
             }

```

Reply:
We're still kinda using:
1) internal property;
2) the fact that all layers are added as persistent layers (there are other methods to add layers);

But it's still better because it will most likely keep us from forgetting to apply this factor to a new layer.

Reply:
Fair enough!
@dzinad , but I'm still wondering if we need to check that sources were set? Do you know if maps has some mechanisms to check if all layers has required sources? Is that important to check them at all? I added the checks just because they were in the original test.